openwisp_controller/config/admin.py

```diff
@@ -559,28 +569,49 @@ def construct_change_message(self, request, form, formsets, add=False):
         return super().construct_change_message(request, form, formsets, add)
 
     def change_group(self, request, queryset):
+        # Validate all selected devices belong to the same organization
+        # which is managed by the user.
+        org_id = None
+        if queryset:
+            org_id = queryset[0].organization_id
+        if (
+            not request.user.is_superuser
+            and str(org_id) not in request.user.organizations_managed
+        ):
+            logger.warning(f'{request.user} does not manage "{org_id}" organization.')
+            return HttpResponseForbidden()
+        if not all(device.organization_id == org_id for device in queryset):
```

Comment:
I am worried this may be problematic if the queryset contains a large number of devices.
Can't we just ensure the queryset filters by the organization_id? Hence we exclude any attempt at tampering with data.

Eg:

```python
queryset = queryset.filter(organization_id=org_id)
```